libs/execution_context/ebpf_native.c

```diff
@@ -0,0 +1,1219 @@
+                                   0xa6f2,
+                                   0x4b53,
+                                   {0xae, 0xa3, 0xf4, 0xa9, 0x4b, 0x9a, 0x80, 0xcb}};
+static ebpf_extension_provider_t* _ebpf_native_provider = NULL;
+
+#define EBPF_CLIENT_TABLE_BUCKET_COUNT 64
+static ebpf_lock_t _ebpf_native_client_table_lock = {0};
+static _Requires_lock_held_(&_ebpf_native_client_table_lock) ebpf_hash_table_t* _ebpf_native_client_table = NULL;
+
+ebpf_result_t
+ebpf_native_load_driver(_In_z_ const wchar_t* service_name);
+void
+ebpf_native_unload_driver(_In_z_ const wchar_t* service_name);
+
+static void
+_ebpf_native_cleanup_maps(_In_reads_(map_count) _Frees_ptr_ ebpf_native_map_t* maps, size_t map_count)
+{
+    for (uint32_t count = 0; count < map_count; count++) {
+        ebpf_native_map_t* map = &maps[count];
+        if (map->pin_path.value && map->pinned && !map->reused) {
+            ebpf_core_update_pinning(UINT64_MAX, &map->pin_path);
+        }
+        if (map->pin_path.value) {
+#pragma warning(push)
+#pragma warning(disable : 6001) // Using uninitialized memory '*maps.pin_path.value'
+            ebpf_free(map->pin_path.value);
+#pragma warning(pop)
+        }
+        if (map->handle != ebpf_handle_invalid) {
+            ebpf_handle_close(map->handle);
+        }
+    }
+
+    ebpf_free(maps);
+}
+
+static void
+_ebpf_native_cleanup_programs(_In_ ebpf_native_program_t* programs, size_t count_of_programs)
+{
+    for (uint32_t i = 0; i < count_of_programs; i++) {
+        if (programs[i].handle != ebpf_handle_invalid) {
+            ebpf_handle_close(programs[i].handle);
+        }
+    }
+
+    ebpf_free(programs);
+}
+
+static void
+_ebpf_native_cleanup_module(_In_ ebpf_native_t* native_module)
+{
+    _ebpf_native_cleanup_maps(native_module->maps, native_module->map_count);
+    _ebpf_native_cleanup_programs(native_module->programs, native_module->program_count);
+
+    ebpf_free(native_module->service_name);
+    ebpf_lock_destroy(&native_module->lock);
+}
+
+void
+ebpf_native_acquire_reference(_In_ ebpf_native_t* native_module)
+{
+    ebpf_assert(native_module->reference_count != 0);
+    ebpf_interlocked_increment_int32(&native_module->reference_count);
+}
+
+void
+ebpf_native_release_reference(_In_opt_ ebpf_native_t* native_module)
+{
+    uint32_t new_ref_count;
+    GUID* module_id = NULL;
+    ebpf_result_t result = EBPF_SUCCESS;
+    bool lock_acquired = false;
+    ebpf_lock_state_t module_lock_state = 0;
+
+    if (!native_module)
+        return;
+
+    ebpf_assert(native_module->reference_count != 0);
+
+    new_ref_count = ebpf_interlocked_decrement_int32(&native_module->reference_count);
+
+    if (new_ref_count == 1) {
+        // Check if all the program references have been released. If that
+        // is the case, explicitly unload the driver, if it is safe to do so.
+        if (ebpf_is_preemptible_work_item_supported()) {
+            bool unload = false;
```

Comment:
Maybe add tracing for this case as it seems complex.

Reply:
Added tracing in the new code added in EC.